src/goto-instrument/contracts/dynamic-frames/dfcc.cpp

```diff
@@ -314,62 +316,73 @@ void dfcct::wrap_replaced_functions()
-        found->first == found->second,
+        !to_check.has_value() || to_check.value().first != str,
         "Function '" + str +
-          "' used as contract for function pointer already the object of a "
-          "contract replacement with '" +
-          id2string(found->second) + "'");
-      log.status() << "Function pointer contract '" << fp_contract
-                   << "' already wrapped with itself in REPLACE mode"
-                   << messaget::eom;
-    }
-    else
-    {
-      // we need to swap it with itself
-      PRECONDITION_WITH_DIAGNOSTICS(
-        utils.function_symbol_exists(str),
-        "Function pointer contract '" + str + "' not found.");
-
-      // triggers signature compatibility checking
-      contract_handler.get_pure_contract_symbol(str);
-
-      log.status() << "Wrapping function pointer contract '" << fp_contract
-                   << "' with itself in REPLACE mode" << messaget::eom;
-
-      swap_and_wrap.swap_and_wrap_replace(
-        fp_contract, fp_contract, function_pointer_contracts);
-      // remove it from the set of symbols to process
-      if(other_symbols.find(fp_contract) != other_symbols.end())
-        other_symbols.erase(fp_contract);
+          "' used as contract for function pointer cannot be itself the object "
+          "of a contract check.");
+
+      // Is it already swapped with another function for contract checking ?
+      auto found = to_replace.find(str);
+      if(found != to_replace.end())
+      {
+        PRECONDITION_WITH_DIAGNOSTICS(
+          found->first == found->second,
+          "Function '" + str +
+            "' used as contract for function pointer already the object of a "
+            "contract replacement with '" +
+            id2string(found->second) + "'");
+        log.status() << "Function pointer contract '" << fp_contract
+                     << "' already wrapped with itself in REPLACE mode"
+                     << messaget::eom;
+      }
+      else
+      {
+        // we need to swap it with itself
+        PRECONDITION_WITH_DIAGNOSTICS(
+          utils.function_symbol_exists(str),
+          "Function pointer contract '" + str + "' not found.");
+
+        // triggers signature compatibility checking
+        contract_handler.get_pure_contract_symbol(str);
+
+        log.status() << "Wrapping function pointer contract '" << fp_contract
+                     << "' with itself in REPLACE mode" << messaget::eom;
+
+        swap_and_wrap.swap_and_wrap_replace(
+          fp_contract, fp_contract, new_contracts);
+        swapped.insert(fp_contract);
+
+        // remove it from the set of symbols to process
+        if(other_symbols.find(fp_contract) != other_symbols.end())
+          other_symbols.erase(fp_contract);
+      }
     }
+    // process newly discovered contracts
+    function_pointer_contracts.clear();
+    function_pointer_contracts.insert(
+      new_contracts.begin(), new_contracts.end());
```

Comment:
Can you just do `function_pointer_contracts = new_contracts;`?

Reply:
yes